### PR TITLE
Support for large numbers of args/kws with Python 3.10

### LINKED ISSUE #7812: Providing many keyword arguments triggers assertion error with Python 3.10 due to unsupported bytecode
It seems that Python 3.10 produces a different bytecode instruction, `CALL_FUNCTION_EX` when using a large number of keyword arguments. From my testing it seems like this occurs once you reach 16 keyword arguments. I only see this behavior on Python 3.10 and not Python 3.9. Here is a reproducer:

```python
import numba

@numba.njit
def my_func1(a=None, b=None, c=None, d=None, e=None, f=None, g=None, h=None, i=None, j=None, k=None, l=None, m=None, n=None, o=None, p=None):
    return my_func2(
        a=a,
        b=b,
        c=c,
        d=d,
        e=e,
        f=f,
        g=g,
        h=h,
        i=i,
        j=j,
        k=k,
        l=l,
        m=m,
        n=n,
        o=o,
        p=p
    )


@numba.njit
def my_func2(a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p):
    return


print(my_func1())
```

### PR DESCRIPTION
**Pull request by njriasan**
Closes #7812. This seems to be working for the set of tests I have covered. I think the code quality probably needs to improve considerably. In addition any code to test for unintended side effects is greatly appreciated.

### CODE REVIEW

**Review by benwilliamgraham — DISMISSED** [review]
Looks good!

File: numba/core/interpreter.py [hunk 1]
```diff
@@ -79,6 +79,432 @@ def get_assignment_source(self, destname):
...
+                        and keyword_def.value.op == "build_map"
+                    )
+                ):
+                    # If we couldn't find where the kwargs are created
+                    # then it should be a normal **kwargs call
+                    # so we produce an unsupported message.
+                    raise UnsupportedError(errmsg)
+                # Determine the kws
+                if keyword_def.value.items:
+                    # n_kws <= 15 case.
```
Comment (ehsantn) [inline numba/core/interpreter.py:442]: I think it would be helpful to add example IRs in comments for each case that is being matched here.

File: numba/core/interpreter.py [hunk 2]
```diff
@@ -79,6 +79,432 @@ def get_assignment_source(self, destname):
...
+                        and args_def.value.op == "build_tuple"
+                    ):
+                        # n_args <= 30 case.
+                        args = _call_function_ex_replace_args_small(
+                            args_def.value,
+                            new_body,
+                            vararg_loc
+                        )
+                    else:
+                        args = _call_function_ex_replace_args_large(
```
Comment (ehsantn) [inline numba/core/interpreter.py:491]: Add comment on the expected IR node for args_def.

File: numba/core/interpreter.py [hunk 3]
```diff
@@ -79,6 +79,432 @@ def get_assignment_source(self, destname):
...
+    n_kws <= 15 and the bytecode looks like:
+
+        BUILD_MAP # Construct the map
+        # Start for each argument
+        LOAD_CONST # Load a constant for the name of the argument
+        LOAD_FAST  # Load each argument.
+        MAP_ADD # Append the (key, value) pair to the map
+        # End for each argument
+
+    In the IR generated, the inital build map is empty and a series
```
Comment (ehsantn) [inline numba/core/interpreter.py:127]:
```suggestion
    In the IR generated, the initial build map is empty and a series
```

**Review by ehsantn — COMMENTED** [review]
Thanks for the effort @njriasan ! I took a quick look and the code makes sense to me.

File: numba/core/interpreter.py (hunk 1 above)
Comment (njriasan) [reply numba/core/interpreter.py:442]: Done!

File: numba/core/interpreter.py (hunk 2 above)
Comment (njriasan) [reply numba/core/interpreter.py:547]: I added a comment explaining my assumptions. Please let me know if it needs further clarification.

**Comment by njriasan** [comment]
@ehsantn and @benwilliamgraham thanks for doing an initial review.

@stuartarchibald could you mark this PR as ready for review? Thanks.

File: numba/core/interpreter.py [hunk 4]
```diff
@@ -79,6 +79,488 @@ def get_assignment_source(self, destname):
...
+                and getattr_stmt.value.op == "getattr"
+                and (
+                    getattr_stmt.value.value.name
+                    == buildmap_name
+                )
+                and getattr_stmt.value.attr == "__setitem__"
+            ):
+                not_found_getattr = False
+            else:
+                search_start += 1
```
Comment (sklam) [inline numba/core/interpreter.py:180]: Why is this needed and what would trigger this line?

File: numba/core/interpreter.py [hunk 5]
```diff
@@ -79,6 +79,488 @@ def get_assignment_source(self, destname):
...
+    if (
+        isinstance(vararg_stmt, ir.Assign)
+        and isinstance(vararg_stmt.value, ir.Var)
+    ):
+        target_name = vararg_stmt.value.name
+        # If there is an initial assignment, delete it
+        new_body[search_end] = None
+        search_end -= 1
+    else:
+        target_name = vararg_stmt.target.name
```
Comment (sklam) [inline numba/core/interpreter.py:288]: Why is this needed and what would trigger this line?

File: numba/core/interpreter.py [hunk 6]
```diff
@@ -79,6 +79,488 @@ def get_assignment_source(self, destname):
...
+                and (
+                    arg_tuple_stmt.value.op
+                    == "build_tuple"
+                )
+                and len(arg_tuple_stmt.value.items) == 1
+            ):
+                # We cannot handle this format.
+                raise UnsupportedError(errmsg)
+            if arg_tuple_stmt.target.name == lhs_name:
+                target_name = rhs_name
```
Comment (sklam) [inline numba/core/interpreter.py:341]: Impossible case? cpython doesn't seem to put the tuple on the LHS.

File: numba/core/interpreter.py [hunk 7]
```diff
@@ -79,6 +79,488 @@ def get_assignment_source(self, destname):
...
+                if (
+                    isinstance(next_stmt, ir.Assign)
+                    and (
+                        next_stmt.target.name
+                        == target_name
+                    )
+                ):
+                    keep_looking = False
+                else:
+                    search_end -= 1
```
Comment (sklam) [inline numba/core/interpreter.py:366]: Why is this needed and what would trigger this line?

**Review by sklam — COMMENTED** [review]
I did a quick check on coverage and found the following lines to be never covered. I only checked a random 20% of the test suite and all the new tests thus the coverage data may not be 100% accurate. If these lines are not really dead, they should have more comments to explain them.

File: numba/core/interpreter.py (hunk 4 above)
Comment (njriasan) [reply numba/core/interpreter.py:180]: This is necessary if some intermediate operation is needed to create the argument. For example, if one of the args was `arg5 = bool(n)`, then search_start skips over the function call.

I can update the comment and the tests to include a function call.

File: numba/core/interpreter.py (hunk 5 above)
Comment (njriasan) [reply numba/core/interpreter.py:288]: This may never be needed, but I wasn't sure the code will always produce an assignment (although I can't construct a case where it hasn't happened).

Basically the existing IR looks like:

```python
...
$final_var = build_tuple(items=[Var(argn, test.py:6)])
$final_tuple = $prev_tuple + $final_var
$varargs_var = $final_tuple
```
but it would be completely valid if instead the IR generated was
```python
...
$final_var = build_tuple(items=[Var(argn, test.py:6)])
$varargs_var = $prev_tuple + $final_var
```
so I was trying to protect against this ever changing.

Comment (njriasan) [reply numba/core/interpreter.py:288]: If you are confident this extra assignment can't occur I'm happy to remove it.

File: numba/core/interpreter.py (hunk 6 above)
Comment (njriasan) [reply numba/core/interpreter.py:341]: I agree I also never saw this on the LHS. I was simply trying to be conservative against any slight changes in the generated code. Since this is a property of the bytecode should I just remove this case?

File: numba/core/interpreter.py (hunk 7 above)
Comment (njriasan) [reply numba/core/interpreter.py:366]: Same reason as the KWS case. If there is a function call like `bool(n)`

Comment (njriasan) [reply numba/core/interpreter.py:366]: If there is anything need to "create" the arg, for example a function call.

**Comment by njriasan** [comment]
> I did a quick check on coverage and found the following lines to be never covered. I only checked a random 20% of the test suite and all the new tests thus the coverage data may not be 100% accurate. If these lines are not really dead, they should have more comments to explain them.

Thanks @sklam. I'll update my tests to add checks to two of the regions and I'll expand the comments on those regions.

Two of the areas that are uncovered were my attempts to be conservative against minor changes in CPython or the IR generated because I wasn't sure if my tests are exhaustive. If you feel like they are unnecessary I'm happy to remove them.

**Comment by njriasan** [comment]
> > I did a quick check on coverage and found the following lines to be never covered. I only checked a random 20% of the test suite and all the new tests thus the coverage data may not be 100% accurate. If these lines are not really dead, they should have more comments to explain them.
> 
> Thanks @sklam. I'll update my tests to add checks to two of the regions and I'll expand the comments on those regions.
> 
> Two of the areas that are uncovered were my attempts to be conservative against minor changes in CPython or the IR generated because I wasn't sure if my tests are exhaustive. If you feel like they are unnecessary I'm happy to remove them.

@sklam I pushed an update to the tests and updated my comments.

File: numba/core/interpreter.py (hunk 5 above)
Comment (sklam) [reply numba/core/interpreter.py:304]: The extra assignment is coming from https://github.com/numba/numba/blob/59fa2e335be68148b3bd72a29de3ff011430038d/numba/core/interpreter.py#L259-L260. 

Let's make this branch raise an error to indicate it is unreachable and add a comment using your description in https://github.com/numba/numba/pull/7866#discussion_r817233996.

File: numba/core/interpreter.py (hunk 6 above)
Comment (sklam) [reply numba/core/interpreter.py:357]: Let's make this raise an error indicating it is unreachable/impossible.

**Comment by stuartarchibald** [comment]
I don't think at present the transform will be able to handle inline control flow at the call site. For example, the rather contrived:
```python
import numba
import dis

@numba.njit
def my_func1(a=None, b=None, c=None, d=None, e=None, f=None, g=None, h=None, i=None, j=None, k=None, l=None, m=None, n=None, o=None, p=None):
    return my_func2(
        a=a,
        b=b,
        c=c,
        d=d,
        e=e,
        f=f,
        g=g,
        h=h,
        i=i,
        j=j,
        k=k,
        l=l,
        m=m,
        n=n,
        o=o,
        p = False if a else True
    )


@numba.njit
def my_func2(a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p):
    return

print(dis.dis(my_func1.py_func))
print(my_func1())
```
which gives a bytecode that has:
```
  6          96 LOAD_CONST              16 ('p')

 22          98 LOAD_FAST                0 (a)
            100 POP_JUMP_IF_FALSE       55 (to 110)
            102 LOAD_CONST              17 (False)
            104 MAP_ADD                  1

  6         106 CALL_FUNCTION_EX         1
            108 RETURN_VALUE

 22     >>  110 LOAD_CONST              18 (True)
            112 MAP_ADD                  1

  6         114 CALL_FUNCTION_EX         1
```
at the location where `p` is resolved and added to the map prior to the call.

**Comment by stuartarchibald** [comment]
> @stuartarchibald Thank you for finding these cases! You are definitely correct and I need to work on a more complete solution.

No problem.

I think given the rarity of the occurrence of these patterns, it would acceptable to do the following:
* Update this patch to catch the cases where control flow occurs. Inform users that control flow on arguments within call sites is not supported and refer them to docs on how to adjust their code.
* Add tests for the above. 

The most common use cases of this somewhat rare pattern would then be covered and workarounds for the uncommon use cases of this rare patter would be shown to users. Essentially, this patch makes the rare patterns work much more often, so I think it's better that it is reviewed and merged with acknowledgement of the imperfection instead of trying to cover everything. This of course also doesn't preclude further fixes later. What do you think?

**Comment by njriasan** [comment]
> > @stuartarchibald Thank you for finding these cases! You are definitely correct and I need to work on a more complete solution.
> 
> No problem.
> 
> I think given the rarity of the occurrence of these patterns, it would acceptable to do the following:
> 
> * Update this patch to catch the cases where control flow occurs. Inform users that control flow on arguments within call sites is not supported and refer them to docs on how to adjust their code.
> * Add tests for the above.
> 
> The most common use cases of this somewhat rare pattern would then be covered and workarounds for the uncommon use cases of this rare pattern would be shown to users. Essentially, this patch makes the rare patterns work much more often, so I think it's better that it is reviewed and merged with acknowledgement of the imperfection instead of trying to cover everything. This of course also doesn't preclude further fixes later. What do you think?

@stuartarchibald I think this approach is reasonable and I agree with it.

**Comment by njriasan** [comment]
@sklam I made your requested changes.

@stuartarchibald I made your changes as well and added tests. Proper error handling requires some extra code, so there may need to be to be a more detailed review. I also added a test for objmode which fails when the definitions are not properly updated.

File: numba/core/interpreter.py [hunk 8]
```diff
@@ -79,6 +79,647 @@ def get_assignment_source(self, destname):
...
+                not_found = True
+                while vararg_loc >= 0 and not_found:
+                    args_def = blk.body[vararg_loc]
+                    if (
+                        isinstance(args_def, ir.Assign)
+                        and args_def.target.name == vararg.name
+                    ):
+                        not_found = False
+                    else:
+                        vararg_loc -= 1
```
Comment (sklam) [inline numba/core/interpreter.py:704]: (Leaving thought trail in case we need it in the future. This probably doesn't present to be a problem)

With bytecode from:

```python
args = (*(0, 1, 2),)   # the starunpack triggers LIST_TO_TUPLE
y = my_func2(*args)
```
```
 16           0 BUILD_LIST               0
              2 LOAD_CONST               1 ((0, 1, 2))
              4 LIST_EXTEND              1
              6 LIST_TO_TUPLE
              8 STORE_FAST              31 (args)

 17          10 LOAD_GLOBAL              0 (my_func2)
             12 LOAD_FAST               31 (args)
             14 CALL_FUNCTION_EX         0
             16 STORE_FAST              32 (y)
```
the code here scan back from the call to `my_func2` to find the definition of `args`, which is suppose to be a `LIST_TO_TUPLE` opcode, but is replaced by `peep_hole_list_to_tuple`; otherwise, the later check will reject this call.

Comment (sklam) [reply numba/core/interpreter.py:704]: I have also tried to trick this part of the code with the walrus operator but Numba is already miscompiling this on mainline:

```python
my_func2(args:=(*(0, 1, 2),), *args)
```

File: numba/core/interpreter.py [hunk 9]
```diff
@@ -79,7 +79,38 @@ def get_assignment_source(self, destname):
...
+        #   172list_to_tuple.116 = $170list_append.115
+        #
+        # but the definition is:
+        #
+        #   list_to_tuple(info=('$4build_list.1',))
+        #
+        func_ir._definitions[lhs].clear()
+    else:
+        raise UnsupportedError(
+            "Inconsistency found the in definitions trying to remove varkwargs"
```
Comment (sklam) [inline numba/core/interpreter.py:107]: Is the error message still accurate? This function seem to be applied generally not just on varkwargs.

Comment (njriasan) [reply numba/core/interpreter.py]: How is this message? Also thank you for the ping, I missed the original review.

File: numba/core/interpreter.py [hunk 10]
```diff
@@ -104,7 +104,9 @@ def _remove_assignment_definition(old_body, idx, func_ir):
         func_ir._definitions[lhs].clear()
     else:
         raise UnsupportedError(
-            "Inconsistency found the in definitions trying to remove varkwargs"
+            "Inconsistency found in the definitions while electuting"
```
Comment (sklam) [inline numba/core/interpreter.py:107]:
```suggestion
            "Inconsistency found in the definitions while executing"
```

File: numba/core/interpreter.py [hunk 11]
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
...
+you have a large number of kwargs and are using inlined control
+flow. You can resolve this issue by moving the control flow out of
+the function call. For example, if you have
+
+    f(a=1 if flag else 0, ...)
+
+Replace that with
+
+    a_val = 1 if flag else 0
+    f(a=a_val, ...)
```
Comment (stuartarchibald) [inline numba/core/interpreter.py:494]: Perhaps `textwrap.dedent` would let this be indented at a more expected level?

File: numba/core/interpreter.py [hunk 12]
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
...
+                kws = call.kws
+                # We search for replace if a call has either vararg
+                # or varkwarg.
+                vararg = call.vararg
+                varkwarg = call.varkwarg
+                start_search = i - 1
+                # varkwarg should be defined second so we start there.
+                varkwarg_loc = start_search
+                keyword_def = None
+                not_found = True
```
Comment (stuartarchibald) [inline numba/core/interpreter.py:518]:
```suggestion
                found = False
```
(and fix up)?

File: numba/tests/test_interpreter.py [hunk 13]
```diff
@@ -0,0 +1,641 @@
...
+            'You can resolve this issue by moving the control flow out',
+            str(raises.exception)
+        )
+        # Uncomment sections when inlined control flow is
+        # actually supported.
+        # py_func = inline_func
+        # cfunc = njit()(inline_func)
+        # a = py_func(False)
+        # b = cfunc(False)
+        # self.assertEqual(a, b)
```
Comment (stuartarchibald) [inline numba/tests/test_interpreter.py:396]: Can probably just remove this, it's easy to add in if/when needed.

File: numba/tests/test_interpreter.py [hunk 14]
```diff
@@ -0,0 +1,641 @@
...
```
Comment (stuartarchibald) [inline numba/tests/test_interpreter.py:513]: Remove? Same reasoning as above.

File: numba/tests/test_interpreter.py [hunk 15]
```diff
@@ -0,0 +1,641 @@
...
+
+    @skip_unless_py10
+    def test_large_args_uninlined_controlflow(self):
```
Comment (stuartarchibald) [inline numba/tests/test_interpreter.py:399]:
```suggestion
    def test_large_args_noninlined_controlflow(self):
```
(suggest same change throughout).

File: numba/tests/test_interpreter.py [hunk 16]
```diff
@@ -0,0 +1,641 @@
...
+            'You can resolve this issue by moving the control flow out',
+            str(raises.exception)
+        )
+        # Uncomment sections when inlined control flow is
+        # actually supported.
+        # py_func = inline_func
+        # cfunc = njit()(inline_func)
+        # a = py_func(False)
+        # b = cfunc(False)
+        # self.assertEqual(a, b)
```
Comment (stuartarchibald) [inline numba/tests/test_interpreter.py:607]: Remove?

**Review by stuartarchibald — COMMENTED** [review]
I've re-reviewed this patch following the changes to catch problems associated with control flow at call sites. It now seems to be handling the cases outlined in the original ticket as expected and failing in a controlled manner else where. Given @sklam has run coverage on this patch I've not checked this further. There's a few minor things to look at in line and once resolved I think this should be ready to merge. Thanks again for your efforts on this @njriasan.

File: numba/core/interpreter.py [hunk 17]
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
...
+    return kws
+
+
+def _call_function_ex_replace_kws_large(
+    old_body, buildmap_name, buildmap_idx, search_end, new_body, func_ir, errmsg
+):
+    """
+    Extracts the kws args passed as varkwarg
+    for CALL_FUNCTION_EX. This pass is taken when
+    n_kws <= 15 and the bytecode looks like:
```
[the remark this answers is not in the document]
Comment (njriasan) [reply numba/core/interpreter.py:153]: This is an error in the docstring. Thank you for catching this.

File: numba/core/interpreter.py [hunk 18]
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
...
+
+        BUILD_MAP # Construct the map
+        # Start for each argument
+        LOAD_CONST # Load a constant for the name of the argument
+        LOAD_FAST  # Load each argument.
+        MAP_ADD # Append the (key, value) pair to the map
+        # End for each argument
+
+    In the IR generated, the initial build map is empty and a series
+    of setitems are applied afterwards. THE IR looks like:
```
[the remark this answers is not in the document]
Comment (njriasan) [reply numba/core/interpreter.py:163]: You are correct. I will update the description.

File: numba/core/interpreter.py [hunk 19]
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
...
+            and isinstance(const_stmt.value, ir.Const)
+        ):
+            # We cannot handle this format so raise the
+            # original error message.
+            raise UnsupportedError(errmsg)
+        key_var_name = const_stmt.target.name
+        key_val = const_stmt.value.value
+        search_start += 1
+        # Now we need to search for a getattr with setitem
+        not_found_getattr = True
```
[the remark this answers is not in the document]
Comment (njriasan) [reply numba/core/interpreter.py:198]: Agreed.

File: numba/core/interpreter.py [hunk 20]
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
...
+            )
+            and len(setitem_stmt.value.args) == 2
+            and (
+                setitem_stmt.value.args[0].name
+                == key_var_name
+            )
+        ):
+            # We cannot handle this format so raise the
+            # original error message.
+            raise UnsupportedError(errmsg)
```
[the remark this answers is not in the document]
Comment (njriasan) [reply numba/core/interpreter.py:257]: Actually for now I will just add the comment. If you would like me to change to an AssertionError I would be happy to, but I'm not certain this is actually unreachable.

File: numba/core/interpreter.py [hunk 21]
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
...
+    look like:
+
+        $orig_tuple = build_tuple(items=[])
+        $first_var = build_tuple(items=[Var(arg0, test.py:6)])
+        $next_tuple = $orig_tuple + $first_var
+        ...
+        $final_var = build_tuple(items=[Var(argn, test.py:6)])
+        $final_tuple = $prev_tuple + $final_var
+        $varargs_var = $final_tuple
+
```
[the remark this answers is not in the document]
Comment (njriasan) [reply numba/core/interpreter.py:328]: Yes this does. I will update the comment.

File: numba/core/interpreter.py [hunk 22]
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
...
+        if (
+            isinstance(concat_stmt, ir.Assign)
+            and concat_stmt.target.name == target_name
+            and isinstance(concat_stmt.value, ir.Expr)
+            and concat_stmt.value.op == "build_tuple"
+            and not concat_stmt.value.items
+        ):
+            # If we have reached the build_tuple
+            # we exit.
+            start_not_found = False
```
[the remark this answers is not in the document]
Comment (njriasan) [reply numba/core/interpreter.py:366]: Yes it is. It is also functioning as a break for the loop, but that can be replaced with a real break.

File: numba/tests/test_interpreter.py (hunk 16 above)
Comment (njriasan) [reply numba/tests/test_interpreter.py]: Done.

File: numba/tests/test_interpreter.py (hunk 15 above)
Comment (njriasan) [reply numba/tests/test_interpreter.py]: Done.

File: numba/tests/test_interpreter.py (hunk 14 above)
Comment (njriasan) [reply numba/tests/test_interpreter.py]: Done.

File: numba/core/interpreter.py (hunk 11 above)
Comment (njriasan) [reply numba/core/interpreter.py]: Agreed. Thank you!

File: numba/core/interpreter.py [hunk 23]
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
...
+    """
+    This peephole rewrites a bytecode sequence unique to Python 3.10
+    where CALL_FUNCTION_EX is used instead of CALL_FUNCTION_KW because of
+    stack limitations set by CPython. This limitation is imposed whenever
+    a function call has too many arguments or keyword arguments.
+
+    https://github.com/python/cpython/blob/a58ebcc701dd6c43630df941481475ff0f615a81/Python/compile.c#L55
+    https://github.com/python/cpython/blob/a58ebcc701dd6c43630df941481475ff0f615a81/Python/compile.c#L4442
+
+    In particular, change is imposed whenever (n_args / 2) + n_kws > 15.
```
[the remark this answers is not in the document]
Comment (njriasan) [reply numba/core/interpreter.py]: Done.

File: numba/core/interpreter.py [hunk 24]
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
...
+        ):
+            # If we have reached the build_tuple
+            # we exit.
+            start_not_found = False
+            new_body[search_end] = None
+            # Remove the definition.
+            _remove_assignment_definition(old_body, search_end, func_ir)
+        else:
+            # We expect to find another arg to append.
+            # The first stmt must be an add
```
[the remark this answers is not in the document]
Comment (njriasan) [reply numba/core/interpreter.py]: Done.

File: numba/core/interpreter.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (njriasan) [reply numba/core/interpreter.py]: Yes.

File: numba/core/interpreter.py [hunk 25]
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
...
+    for CALL_FUNCTION_EX. This pass is taken when
+    n_kws <= 15 and the bytecode looks like:
+
+        # Start for each argument
+        LOAD_FAST  # Load each argument.
+        # End for each argument
+        ...
+        BUILD_CONST_KEY_MAP # Build a map
+
+    In the IR generated, the varkwarg refer
```
[the remark this answers is not in the document]
Comment (njriasan) [reply numba/core/interpreter.py]: Done.

File: numba/core/interpreter.py [hunk 26]
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
...
+
+        $orig_tuple = build_tuple(items=[])
+        $first_var = build_tuple(items=[Var(arg0, test.py:6)])
+        $next_tuple = $orig_tuple + $first_var
+        ...
+        $final_var = build_tuple(items=[Var(argn, test.py:6)])
+        $final_tuple = $prev_tuple + $final_var
+        $varargs_var = $final_tuple
+
+    It unclear if the extra assignment at the end is always present.
```
[the remark this answers is not in the document]
Comment (njriasan) [reply numba/core/interpreter.py]: Yes this included in the AssertionError, so I will remove this from the comments.

File: numba/core/interpreter.py [hunk 27]
```diff
@@ -79,6 +79,649 @@ def get_assignment_source(self, destname):
...
+                and isinstance(stmt.value, ir.Expr)
+                and stmt.value.op == "call"
+                and stmt.value.varkwarg is not None
+            ):
+                blk_changed = True
+                call = stmt.value
+                args = call.args
+                kws = call.kws
+                # We search for replace if a call has either vararg
+                # or varkwarg.
```
[the remark this answers is not in the document]
Comment (njriasan) [reply numba/core/interpreter.py]: I have updated this with a longer comment explaining the meaning of this. The summary is that we investigate a call expression whenever it contains either vararg or varkwarg. varkwarg requires the bytecode changes and vararg is used for generating a nicer error message.

Please let me know if you think my more detailed comment is still unclear.

**Comment by njriasan** [comment]
@stuartarchibald I made your requested changes. There are possibly two things that need to be looked at:

1. Should the "abundance of caution" check contain a different error like an unreachable assertion. I don't have enough confidence with the code structure to say its truly unreachable so I just added a comment.
2. Is the comment explaining when to check the call expression clear enough?

If both these changes seem fine, I think everything else should be a direct application of your feedback and we should be ready to merge. Thank you so much for the detailed review!

File: numba/core/interpreter.py [hunk 28]
```diff
@@ -79,6 +80,646 @@ def get_assignment_source(self, destname):
...
+                isinstance(stmt, ir.Assign)
+                and isinstance(stmt.value, ir.Expr)
+                and stmt.value.op == "call"
+                and stmt.value.varkwarg is not None
+            ):
+                blk_changed = True
+                call = stmt.value
+                args = call.args
+                kws = call.kws
+                # We need to to check the call expression contents if
```
Comment (stuartarchibald) [inline numba/core/interpreter.py:507]:
```suggestion
                # We need to check the call expression contents if
```

File: numba/core/interpreter.py [hunk 29]
```diff
@@ -79,6 +80,646 @@ def get_assignment_source(self, destname):
...
+                # it contains either vararg or varkwarg. If it contains
+                # varkwarg we need to update the actual bytecode. If it
+                # just contains vararg we don't need to update the bytecode,
```
Comment (stuartarchibald) [inline numba/core/interpreter.py:510]: Do you mean `bytecode` or IR?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the fixes @njriasan, there's a typo and a query to resolve else this looks good.

**Comment by stuartarchibald** [comment]
> @stuartarchibald I made your requested changes.

@njriasan Many thanks.

> There are possibly two things that need to be looked at:
> 
>     1. Should the "abundance of caution" check contain a different error like an unreachable assertion. I don't have enough confidence with the code structure to say its truly unreachable so I just added a comment.

I think it can be left as-is for now. Given it cannot be handled at present, if it turns out that there's use cases in practice that will trigger it, I would guess someone will make a request for it/report it. I don't think it warrants an exception containing the reportable condition text because it's already known that it might happen and it can't be handled.
 
>     2. Is the comment explaining when to check the call expression clear enough?

Couple of minor things I've noted in https://github.com/numba/numba/pull/7866#pullrequestreview-964652343 else looks good thanks.

> If both these changes seem fine, I think everything else should be a direct application of your feedback and we should be ready to merge. Thank you so much for the detailed review!

No problem, thanks for your efforts on this.

File: numba/core/interpreter.py (hunk 28 above)
Comment (njriasan) [reply numba/core/interpreter.py]: Done.

**Comment by njriasan** [comment]
Thank you @stuartarchibald. I made the requested changes.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes @njriasan, great to get this merged!
